Next RPC Indication for the HMI (#824)
smartdevicelink/sdl_evolution: proposals/NNNN-NextRpcIndication.md

# Next RPC Indication for the HMI

* Proposal: [SDL-NNNN](NNNN-NextRPCIndication.md)

> * SDL Proposal: [To be filled later](https://github.com/smartdevicelink/sdl_evolution/blob/master/proposals/)

Reviewer: This is not part of the SDL Evolution Proposal template - please remove

* Author: [Michael Crimando](https://github.com/MichaelCrimando)
* Status: **Awaiting review**
* Impacted Platforms: [Core / iOS / Android / Web / RPC ]

## Introduction

In situations with driver distraction, also with making a good UI, it really hurts the possible UX when the HMI has no idea what's going to happen next after the user taps an in-app menu item, a softbutton or a selection in a perform interaction. 

> This feature would provide Apps a way for the HMI to know what's next when a button or choice is selected.

Reviewer: ```suggestion
This feature would provide apps with a way for the HMI to know what's next when a button or choice is selected.
```

## Motivation

As a system, I want to be able to be smarter and faster with in-app menu systems.
This enables:

> 1. Better HMI possible. Can show forward and backwards arrows for the user so the user knows what to expect when navigating around.

Reviewer: ```suggestion
1. **Better HMI possibilities:** can show forward and backward arrows for the user so the user knows what to expect when navigating around.
```

> 2. Smart Driver Distraction. If a softbutton or choice would lead to a screen that's normally locked out while driving (like a Scrollable Message), the HMI could show the softbutton as disable while driving.

Reviewer: ```suggestion
2. **Smart Driver Distraction:** if a softbutton or choice would lead to a screen that's normally locked out while driving (like a Scrollable Message), the HMI could show the softbutton as disabled while driving.
```

> 3. The HMI can have a much smoother experience with native HMI loading screens - not depending on the mobile app to send a sort of loading screen on it's own.

Reviewer: ```suggestion
3. The HMI can have a much smoother experience with native HMI loading screens - not depending on the mobile app to send a sort of loading screen on its own.
```

Here's an example of current SYNC3 iHeartRadio flow (left) vs potential future (right).
With this proposal, the HMI would know that after the user picks something like "Radio Genres", another RPC is incoming and can show a special loading screen until the RPC comes in. (Or the loading screen times out, but that behavior would be set by the OEM and not require SDL responses of any sort)
![App screen and display](../assets/proposals/NNNN-NextRpcIndication/NNNN-NextRpcIndication.png)

## Proposed solution

Add a new struct `NextFunctionInfo` with all the data needed for the HMI to know what's coming next to the HMI API and the same in the MOBILE API:
```xml
<struct name="NextFunctionInfo" since="x.x">
	<description>
		Outlines information about the next RPC that will be triggered.		
	</description>
	<param name="nextFunctionID" type="FunctionID" mandatory="true"/>

> 		<description>The next Function (RPC) that will be triggered by the selecting the current option/command/choice etc.</description>

Reviewer: ```suggestion
		<description>The next function (RPC) that will be triggered by selecting the current option/command/choice etc.</description>
```

	</param>
		<param name="nextInitialText" type="String" maxlength="500"  mandatory="false"/>
			<description>This lets the HMI know what the initialText of the next RPC is going to be to help the loading user experience.</description>
		</param>
</struct>
```

Add the struct to `Choice` in the HMI API and the same in the MOBILE API:
```xml
<struct name="Choice" since="1.0">
  <description>A choice is an option given to the user, which can be selected either by menu, or through voice recognition system.</description>
  .
  .
  .
  <param name="nextFunctionInfo" type="NextFunctionInfo" mandatory="false" since="x.x">

>     <description>Information about the next RPC that is triggered when this choice is selected </description>

Reviewer: ```suggestion
    <description>Information about the next RPC that is triggered when this choice is selected. </description>
```

  </param>
</struct>
```

> Add the struct to `AddCommand` to the HMI API

Reviewer: ```suggestion
Add the struct to `AddCommand` to the HMI API:
```

```xml
  <function name="AddCommand" functionID="AddCommandID" messagetype="request" since="1.0">
  <description>
    Adds a command to the in application menu.
    Either menuParams or vrCommands must be provided.
  </description>
  .
  .
  .
  <param name="nextFunctionInfo" type="NextFunctionInfo" mandatory="false" since="x.x">
    <description>

>       Information about the next RPC that is triggered when this choice is selected

Reviewer: ```suggestion
      Information about the next RPC that is triggered when this choice is selected.
```

    </description>
  </param>
</function>
```

Add the struct to the `SoftButton` struct. This will cover Show/SetDisplayLayout, Alert, ShowconstantTBT and ScrollableMessage in the HMI API and the same in the MOBILE API
```xml
<struct name="SoftButton">
  .
  .
  .
  <param name="nextFunctionInfo" type="NextFunctionInfo" mandatory="false" since="x.x">

>     <description>Information about the next RPC that is triggered when this choice is selected </description>

Reviewer: ```suggestion
    <description>Information about the next RPC that is triggered when this choice is selected. </description>
```
